File: src/agents/compact.ts
import { estimateTokens } from '../util/text.ts';
import type { SubagentResult } from './types.ts';

/**
 * Context management (subagent coordinator).
 *
 * The coordinator only ever passes compact summaries between steps — never
 * raw tool dumps or full conversations. `compactResults` shrinks the
 * accumulated results to a token budget by truncating prose and dropping the
 * oldest results first, so subagents stay small and TPM limits are never hit.
 */

const SUMMARY_CHARS = 600;
const FINDING_CHARS = 220;

function truncate(text: string, max: number): string {
  if (text.length <= max) return text;
  // Avoid splitting a UTF-16 surrogate pair (no lone surrogates in output).
  let cut = max - 1;
  const code = text.charCodeAt(cut);
  if (code >= 0xdc00 && code <= 0xdfff) cut -= 1;
  return text.slice(0, Math.max(0, cut)) + '…';
}

/** Render one result as a compact context block. */
export function renderResult(result: SubagentResult, budgetChars: number): string {
  const lines: string[] = [];
  lines.push(`- [${result.label}${result.status === 'failed' ? ' (failed)' : ''}] ${truncate(result.summary, budgetChars)}`);
  if (result.files.length > 0) lines.push(`  files: ${result.files.slice(0, 8).join(', ')}`);
  for (const f of result.findings.slice(0, 3)) lines.push(`  • ${truncate(f, FINDING_CHARS)}`);
  for (const r of result.recommendations.slice(0, 2)) lines.push(`  → ${truncate(r, FINDING_CHARS)}`);
  if (result.error) lines.push(`  error: ${truncate(result.error, 240)}`);
  return lines.join('\n');
}

/**
 * Compact a list of results into one context string under a token budget.
 * Deterministic: newer results are kept, older ones are truncated/dropped.
 */
export function compactResults(results: SubagentResult[], budgetTokens: number): string {
  const budget = Math.max(200, budgetTokens);
  const blocks: string[] = [];
  let used = 0;
  // Walk newest-first so the most relevant (last) results survive compaction.
  for (let i = results.length - 1; i >= 0; i -= 1) {
    const result = results[i] as SubagentResult;
    if (result.status === 'skipped') continue;
    let block = renderResult(result, SUMMARY_CHARS);
    const cost = estimateTokens(block);
    if (used + cost > budget && blocks.length > 0) break; // drop oldest results
    // Degrade the summary further for very large results.
    while (used + estimateTokens(block) > budget && block.length > 160) {
      block = truncate(block, Math.floor(block.length / 2));
    }
    // Hard clamp: even a short block must fit the remaining budget.
    if (used + estimateTokens(block) > budget) {
      block = truncate(block, Math.max(80, budget - used));
    }
    blocks.unshift(block);
    used += estimateTokens(block);
  }
  return blocks.join('\n');
}

/** Compact arbitrary free text to a char cap (used for planner context). */
export function compactText(text: string, maxChars: number): string {
  if (text.length <= maxChars) return text;
  const head = text.slice(0, Math.floor(maxChars * 0.7));
  const tail = text.slice(-Math.floor(maxChars * 0.3));
  return `${head}\n… [context truncated, ${text.length - maxChars} chars omitted] …\n${tail}`;
}

File: grace/agents/compact.py
"""Context management for the coordinator (port of src/agents/compact.ts).

The coordinator only ever passes compact summaries between steps — never raw
tool dumps or full conversations. `compact_results` shrinks the accumulated
results to a token budget by truncating prose and dropping the oldest results
first.
"""

from grace.util_text import estimate_tokens

SUMMARY_CHARS = 600
FINDING_CHARS = 220


def _truncate(text: str, max_chars: int) -> str:
    if len(text) <= max_chars:
        return text
    cut = max_chars - 1
    return text[:max(0, cut)] + "…"


def render_result(result, budget_chars: int) -> str:
    lines = [f"- [{result.label}{' (failed)' if result.status == 'failed' else ''}] {_truncate(result.summary, budget_chars)}"]
    if result.files:
        lines.append(f"  files: {', '.join(result.files[:8])}")
    for f in result.findings[:3]:
        lines.append(f"  • {_truncate(f, FINDING_CHARS)}")
    for r in result.recommendations[:2]:
        lines.append(f"  → {_truncate(r, FINDING_CHARS)}")
    if result.error:
        lines.append(f"  error: {_truncate(result.error, 240)}")
    return "\n".join(lines)


def compact_results(results: list, budget_tokens: int) -> str:
    """Compact a list of results into one context string under a token budget.
    Deterministic: newer results are kept, older ones are truncated/dropped."""
    budget = max(200, budget_tokens)
    blocks: list[str] = []
    used = 0
    # Walk newest-first so the most relevant (last) results survive compaction.
    for result in reversed(results):
        if result.status == "skipped":
            continue
        block = render_result(result, SUMMARY_CHARS)
        cost = estimate_tokens(block)
        if used + cost > budget and blocks:
            break  # drop oldest results
        # Degrade the summary further for very large results.
        while used + estimate_tokens(block) > budget and len(block) > 160:
            block = _truncate(block, len(block) // 2)
        # Hard clamp: even a short block must fit the remaining budget.
        if used + estimate_tokens(block) > budget:
            block = _truncate(block, max(80, budget - used))
        blocks.insert(0, block)
        used += estimate_tokens(block)
    return "\n".join(blocks)


def compact_text(text: str, max_chars: int) -> str:
    """Compact arbitrary free text to a char cap (used for planner context)."""
    if len(text) <= max_chars:
        return text
    head = text[:int(max_chars * 0.7)]
    tail = text[-int(max_chars * 0.3):]
    return f"{head}\n… [context truncated, {len(text) - max_chars} chars omitted] …\n{tail}"
